community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/BoltAuthIT.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.v1.runtime.integration;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.neo4j.bolt.v1.runtime.Session;
+import org.neo4j.kernel.api.exceptions.Status;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.neo4j.bolt.v1.runtime.integration.SessionMatchers.failedWith;
+import static org.neo4j.bolt.v1.runtime.integration.SessionMatchers.recorded;
+import static org.neo4j.bolt.v1.runtime.integration.SessionMatchers.successButRequiresPasswordChange;
+import static org.neo4j.helpers.collection.MapUtil.map;
+
+public class BoltAuthIT
+{
+    @Rule
+    public TestSessions env = new TestSessions().withAuthEnabled( true );
+
+    @Test
+    public void shouldGiveCredentialsExpiredStatusOnExpiredCredentials() throws Throwable
+    {
+        // given it is important for client applications to programmatically
+        // identify expired credentials as the cause of not being authenticated
+        Session session = env.newSession( "test" );
+        RecordingCallback recorder = new RecordingCallback();
+
+        // when
+        session.init( "TestClient/1.0.0", map(
+                "scheme", "basic",
+                "principal", "neo4j",
+                "credentials", "neo4j" ), null, recorder );
+        session.run( "CREATE ()", map(), null, recorder );
```

Comment:
So, I think doing it this way is right, for two reasons; 

one, the way you'll be resetting the password will be through a procedure call, so you need to get past `INIT` successfully to run that procedure call;

two, credentials expiring should override ongoing sessions - if you've got good uptime, you could have an app tier running month-old sessions or longer. If we let credentials expire over time or let an admin expire them, we'd want those sessions to get helpful errors saying to reset the password.

There's a third benefit, which is that we remove a piece of complexity from Bolt and use a pipeline that's already there - the error signaling system - which drops the number of different things an implementor has to learn, and the number of moving parts we need to maintain